From 3d74ac29e7456c77b32949131cf8ad6190d5c460 Mon Sep 17 00:00:00 2001
From: Mooffie <mooffie@gmail.com>
Date: Mon, 19 Dec 2016 02:06:18 +0200
Subject: [PATCH 2/6] Ticket #2707: extfs: u7z: handle missing size.

The uncompressed size may be missing in 7z's output (gods know why). We use the
compressed size in this case.
---
 src/vfs/extfs/helpers/u7z | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/src/vfs/extfs/helpers/u7z b/src/vfs/extfs/helpers/u7z
index 8270a20..52d123d 100755
--- a/src/vfs/extfs/helpers/u7z
+++ b/src/vfs/extfs/helpers/u7z
@@ -26,9 +26,12 @@ mcu7zip_list ()
 
         ugid="`id -nu` `id -ng`"
 
-        date_re='^\(....\)-\(..\)-\(..\) \(..:..:..\)'
+        date_re='^\(....\)-\(..\)-\(..\) \(..:..:..\)' # 19 chars.
         date_mc='\2-\3-\1 \4'
-        size_re='............'
+
+        size_re='............' # 12 chars.
+        empty_size_re=' \{12\}'
+        zero_size='           0'
 
         # archive entries can have no datetime info, 7z will use archive file datetime
         date_archive=`stat -c %y "$1" 2>/dev/null | sed -n "s/${date_re}.*/${date_mc}/p" 2>/dev/null`
@@ -37,6 +40,16 @@ mcu7zip_list ()
 
         $P7ZIP l "$1" | sed -n "
 
+                  # If the uncompressed size is missing, we copy the compressed size onto it.
+                  #
+                  # But first, if the compressed size is missing too, set it to zero:
+                  s/^\(.\{19\} [D.]....\) $empty_size_re $empty_size_re/\1 $zero_size $zero_size/
+                  # Next, do the copy:
+                  s/^\(.\{19\} [D.]....\) $empty_size_re \($size_re\)/\1 \2 \2/
+                  #
+                  # (We use '.\{19\}' as the date may be missing. It may give false positives
+                  # but we don't mind: the printing commands that follow use strict patterns.).
+
                   # Handle directories.
                   s/$date_re D.... $size_re $size_re\(.*\)/drwxr-xr-x 1 $ugid 0 $date_mc \5/p
                        s/^\s*D.... $size_re $size_re\(.*\)/drwxr-xr-x 1 $ugid 0 $date_archive \1/p
-- 
2.9.3

